pkg/router/crl/crl.go

```diff
@@ -209,30 +388,111 @@ func getHTTPCRL(url string) (*pkix.CertificateList, error) {
+	//	   still be in the middle of a CA bundle/CRL refresh, so the CRL management loop might be actively working with
+	//	   this directory
+	// Everything else can be deleted
+	var newestDirName string
+	for _, dirEntry := range dirEntries {
+		entryName := dirEntry.Name()
+		// entryName is the bare file/directory name, but mtlsLatestSymlink, mtlsLatestStagingSymlink, and inUseDirName
+		// are all fully-qualified paths. Since we know that all the dir entries are the files/directories in
+		// mtlsBaseDirectory, we can prepend that to entryName to get the full name for comparisons now, and for
+		// os.RemoveAll() if it needs to be deleted later in this function
+		fullPath := filepath.Join(mtlsBaseDirectory, entryName)
+		log.V(4).Info(fmt.Sprintf("Checking %s (%s)", entryName, fullPath))
+		// Leave mtlsLatestSymlink, mtlsLatestStagingSymlink, and the current directory pointed to by
+		// mtlsLatestSymlink alone.
+		if fullPath == mtlsLatestSymlink || fullPath == mtlsLatestStagingSymlink || fullPath == inUseDirName {
+			continue
+		}
+		dirEntryInfo, err := dirEntry.Info()
 		if err != nil {
-			return crlForKeyId, err
+			continue
 		}
-		for _, ext := range crl.TBSCertList.Extensions {
-			if ext.Id.Equal(authorityKeyIdentifierOID) {
-				var authKeyId authorityKeyIdentifier
-				if _, err := asn1.Unmarshal(ext.Value, &authKeyId); err != nil {
-					return crlForKeyId, err
+		dirUpdateTimestamp := dirEntryInfo.ModTime()
+		if dirUpdateTimestamp.Before(mostRecentTimestamp) {
+			// If this isn't the newest directory and it's not pointed to by mtlsLatestSymlink, we definitely don't need it.
+			log.V(2).Info(fmt.Sprintf("Removing %s because it's stale", entryName))
+			err := os.RemoveAll(filepath.Join(mtlsBaseDirectory, entryName))
+			if err != nil {
+				log.Error(err, fmt.Sprintf("Failed to remove %s", entryName))
+				continue
+			}
+		} else {
+			// If this is newer than the newest directory yet seen, delete the directory we thought was newest, then
+			// keep its name as the current newest directory
+			mostRecentTimestamp = dirUpdateTimestamp
+			if newestDirName != "" {
+				log.V(2).Info(fmt.Sprintf("Removing %s because it's stale", newestDirName))
+				err := os.RemoveAll(filepath.Join(mtlsBaseDirectory, newestDirName))
+				if err != nil {
+					log.Error(err, fmt.Sprintf("Failed to remove %s", newestDirName))
 				}
-				subjectKeyId := hex.EncodeToString(authKeyId.KeyIdentifier)
-				crlForKeyId[subjectKeyId] = crl
 			}
+			newestDirName = entryName
 		}
-		crlData = data
 	}
-	return crlForKeyId, nil
+	log.V(4).Info("Cleanup done")
+	return nil
+}
+
+// newMTLSDirectory creates a new temp directory for writing the CA bundle and CRL file, and returns the full path to
+// that directory. This directory will be cleaned up by the mTLS directory reaper once a newer version is present
+func newMTLSDirectory() (string, error) {
+	subdir := fmt.Sprintf("mtls-%s", time.Now().Format(time.RFC3339))
```

Comment:
If anyone ever tries to port OpenShift router to MS Windows, they will probably be annoyed by this code.  * grin *.  Would `fmt.Sprintf("mtls-%d", time.Now().Unix())` be better?  Or simply use `os.MkdirTemp(mtlsBaseDirectory, "mtls-")`?

Reply:
Thanks for mentioning `os.MkdirTemp`; I wasn't familiar with that, but it fits this use case very well. Using that, `newMTLSDirectory` basically boiled down to `return os.MkdirTemp(...)`, so I removed the function and now it just calls `MkdirTemp` directly

Reply:
After talking with Andy, I switched back to using a date-based ID in the mtls directory names. I think it makes the directory names easier to remember when debugging, and it should still result in unique directory names.